lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp

```diff
@@ -36,17 +41,71 @@ bool LegalGridAnalysis::applyOverrides() {
-  // For now result of analysis is maximum supported grid size.
-  //
+  // A first incomplete implementation of the LegalGridAnalysis.
+  // This implementation is a placeholder and is meant to just enable testing of
+  // other components.
+
+  // Get output tensor type.
+  // TODO: This ignores multiple outputs...?
   RankedTensorType tensorType =
       mlir::cast<RankedTensorType>(op->getResult(0).getType());
   LayoutAttr layout = mlir::cast<LayoutAttr>(tensorType.getEncoding());
-  llvm::ArrayRef<int64_t> tensorShape = tensorType.getShape();
 
-  analysisResult.push_back(layout.withGrid(
-      op->getContext(), tensorShape,
-      GridAttr::get(op->getContext(), analysisInput.maxGrid.getShape())));
+  // DRAM
+  // No grid is set since the tensor is not sharded.
+  // TODO: Is this a viable solution or should we have a grid?
+  LayoutAttr dram =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceDRAM);
+  analysisResult.push_back(dram);
+
+  // L1 Interleaved (same as above)
+  LayoutAttr l1Interleaved =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceL1);
+  analysisResult.push_back(l1Interleaved);
+
+  // L1 Sharded
+  LayoutAttr shardedBase =
+      layout.withMemorySpace(op->getContext(), MemorySpace::DeviceL1);
+
+  // Block Sharded
+  for (auto width = 2; width <= analysisInput.maxGrid.getShape()[0]; ++width) {
+    for (auto height = 2; height <= analysisInput.maxGrid.getShape()[1];
+         ++height) {
+      analysisResult.push_back(shardedBase.withGrid(
+          op->getContext(), tensorType,
+          GridAttr::get(op->getContext(), {width, height})));
+    }
+  }
+
+  auto numCores =
+      analysisInput.maxGrid.getShape()[0] * analysisInput.maxGrid.getShape()[1];
+  // Height Sharded
+  // TODO: Missing affine mapping to actual grid.
+  // TODO: Can we have every shape of 1d grid? Probably not, need to check what
+  // is divisible by grid sides.
+  // TODO: Limit the number of options to some reasonable number.
+  // TODO: Put all of this into the same loop.
+  for (auto height = 2; height <= numCores; ++height) {
+    analysisResult.push_back(
+        shardedBase.withGrid(op->getContext(), tensorType,
+                             GridAttr::get(op->getContext(), {height, 1})));
+  }
+
+  // Width Sharded
+  for (auto width = 2; width <= numCores; ++width) {
+    analysisResult.push_back(
+        shardedBase.withGrid(op->getContext(), tensorType,
+                             GridAttr::get(op->getContext(), {1, width})));
+  }
+
+  // Filter layouts based on output tensor legality for current op.
+  analysisResult.erase(
+      std::remove_if(analysisResult.begin(), analysisResult.end(),
+                     [this](LayoutAttr layout) {
+                       return !mock_is_output_tensor_legal_for_op(op, layout);
+                     }),
+      analysisResult.end());
+
+  // TODO: Potetialy filter out tensors that dont fit into L1 at all.
```

Comment:
It would be nice to have this for sharded ones.

Reply:
Ok, will treat this as a nice to have.

Reply:
Actually, don't do this. :) We can still decide later to split the tensor for pipeline op, so this will be up to sharding policy to decide.